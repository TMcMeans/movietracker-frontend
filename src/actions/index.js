export const addMovies = movieArray => {
  return {
    type: 'ADD_MOVIES',
    movies: movieArray
  };
};

export const loginUser = user => {
  return {
    type: 'LOGIN_USER',
    user
  };
};

export const logoutUser = () => {
  return {
    type: 'LOGOUT_USER'
  };
};

export const hasErrored = message => ({
  type: 'HAS_ERRORED',
  message
});

export const removeFavorites = () => ({
  type: 'REMOVE_FAVORITES'
});

export const retrieveAllFavorites = favorites => {
  return {
    type: 'RETRIEVE_ALL_FAVORITES',
    favorites
  };
};

export const addMessage = message => {
  return {
    type: 'ADD_MESSAGE',
    message
  };
};
