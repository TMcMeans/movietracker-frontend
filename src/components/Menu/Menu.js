import React, { Component } from 'react';
import Login from '../../containers/Login/Login';
import { Link } from 'react-router-dom';

class Menu extends Component {
  constructor() {
    super();
    this.state = {
      fullMenu: false
    };
  }

  toggleMenu = () => {
    this.setState({ fullMenu: !this.state.fullMenu });
  };

  render() {
    if (!this.state.fullMenu) {
      return (
        <div className="menu" onClick={this.toggleMenu}>
          <p>MENU</p>
        </div>
      );
    } else {
      return (
        <div className="full-menu">
          <button onClick={this.toggleMenu}>X</button>
          <p>
            <Link to="/login" onClick={this.toggleMenu}>
              Login
            </Link>
          </p>
          <p>
            <Link to="/favorites" onClick={this.toggleMenu}>
              Stalked:0
            </Link>
          </p>
        </div>
      );
    }
  }
}

export default Menu;
