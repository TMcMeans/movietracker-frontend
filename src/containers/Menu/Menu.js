import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { logoutUser, addMessage } from '../../actions';
import { connect } from 'react-redux';

export class Menu extends Component {
  constructor() {
    super();
    this.state = {
      fullMenu: false
    };
  }

  toggleMenu = () => {
    this.setState({ fullMenu: !this.state.fullMenu });
  };

  handleClick = () => {
    this.props.logoutUser();
    this.props.addMessage('You are now logged out')
    this.toggleMenu();
  };

  render() {
    let log = (
      <p>
        <Link className="menu-item" to="/login" onClick={this.toggleMenu}>
          Login
        </Link>
      </p>
    );
    
    if (this.props.user) {
      log = (
        <p className="menu-item" id="logout" onClick={this.handleClick}>
          Logout
        </p>
      );
    }
    if (!this.state.fullMenu) {
      return (
        <div className="menu" onClick={this.toggleMenu}>
          <p>MENU</p>
        </div>
      );
    } else {
      return (
        <div className="full-menu">
          <button onClick={this.toggleMenu}>X</button>
          {log}
          <p>
            <Link className='menu-item' to="/favorites" onClick={this.toggleMenu}>
              Stalked: <span>{this.props.allFavorites.length}</span>
            </Link>
          </p>
        </div>
      );
    }
  }
}

export const mapStateToProps = (state) => ({
  user: state.user,
  allFavorites: state.favorites
})

export const mapDispatchToProps = dispatch => ({
  logoutUser: () => dispatch(logoutUser()),
  addMessage: (message) => dispatch(addMessage(message))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Menu);

Menu.propTypes = {
  user: PropTypes.object,
  logoutUser: PropTypes.func.isRequired
};
