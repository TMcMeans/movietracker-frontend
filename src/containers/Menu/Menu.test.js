import React from 'react';
import { shallow } from 'enzyme';
import { Menu, mapStateToProps, mapDispatchToProps } from './Menu';

import { logoutUser } from '../../actions';

describe('Menu', () => {
  const mockUser = { id: 1 };
  const mockFunc = jest.fn();
  const wrapper = shallow(<Menu logoutUser={mockFunc} />);
  it('should match snapshot with correct data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should change state of fullMenu to true when toggleMenu is called', () => {
    const expected = {
      fullMenu: true
    };
    wrapper.instance().toggleMenu();
    expect(wrapper.state()).toEqual(expected);
  });

  describe('mapStateToProps', () => {
    it('should return a user object', () => {
      const mockState = { user: mockUser, logoutUser: mockFunc };

      const expected = {
        user: mockUser
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call logoutUser when handleClick is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = logoutUser();

      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.logoutUser();

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
