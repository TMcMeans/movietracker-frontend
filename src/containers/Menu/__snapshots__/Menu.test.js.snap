// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should match snapshot with correct data 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Menu
    addMessage={[MockFunction]}
    allFavorites={Array []}
    logoutUser={[MockFunction]}
    removeFavorites={[MockFunction]}
    user={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        MENU
      </p>,
      "className": "menu",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "MENU",
      },
      "ref": null,
      "rendered": "MENU",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          MENU
        </p>,
        "className": "menu",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "MENU",
        },
        "ref": null,
        "rendered": "MENU",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Menu should match the snapshot when there is a user 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Menu
    addMessage={[MockFunction]}
    allFavorites={Array []}
    logoutUser={[MockFunction]}
    removeFavorites={[MockFunction]}
    user={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p>
          <Link
            className="menu-item a-tag"
            onClick={[Function]}
            replace={false}
            to="/"
          >
            Home
          </Link>
        </p>,
        <button
          onClick={[Function]}
        >
          X
        </button>,
        <p>
          <Link
            className="menu-item"
            onClick={[Function]}
            replace={false}
            to="/login"
          >
            Login
          </Link>
        </p>,
        <p>
          <Link
            className="menu-item a-tag"
            onClick={[Function]}
            replace={false}
            to="/favorites"
          >
            Stalked: 
            <span>
              0
            </span>
          </Link>
        </p>,
      ],
      "className": "full-menu",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Link
            className="menu-item a-tag"
            onClick={[Function]}
            replace={false}
            to="/"
          >
            Home
          </Link>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Home",
            "className": "menu-item a-tag",
            "onClick": [Function],
            "replace": false,
            "to": "/",
          },
          "ref": null,
          "rendered": "Home",
          "type": [Function],
        },
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "X",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "X",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Link
            className="menu-item"
            onClick={[Function]}
            replace={false}
            to="/login"
          >
            Login
          </Link>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Login",
            "className": "menu-item",
            "onClick": [Function],
            "replace": false,
            "to": "/login",
          },
          "ref": null,
          "rendered": "Login",
          "type": [Function],
        },
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Link
            className="menu-item a-tag"
            onClick={[Function]}
            replace={false}
            to="/favorites"
          >
            Stalked: 
            <span>
              0
            </span>
          </Link>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Stalked: ",
              <span>
                0
              </span>,
            ],
            "className": "menu-item a-tag",
            "onClick": [Function],
            "replace": false,
            "to": "/favorites",
          },
          "ref": null,
          "rendered": Array [
            "Stalked: ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 0,
              },
              "ref": null,
              "rendered": 0,
              "type": "span",
            },
          ],
          "type": [Function],
        },
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
            <Link
              className="menu-item a-tag"
              onClick={[Function]}
              replace={false}
              to="/"
            >
              Home
            </Link>
          </p>,
          <button
            onClick={[Function]}
          >
            X
          </button>,
          <p>
            <Link
              className="menu-item"
              onClick={[Function]}
              replace={false}
              to="/login"
            >
              Login
            </Link>
          </p>,
          <p>
            <Link
              className="menu-item a-tag"
              onClick={[Function]}
              replace={false}
              to="/favorites"
            >
              Stalked: 
              <span>
                0
              </span>
            </Link>
          </p>,
        ],
        "className": "full-menu",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Link
              className="menu-item a-tag"
              onClick={[Function]}
              replace={false}
              to="/"
            >
              Home
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Home",
              "className": "menu-item a-tag",
              "onClick": [Function],
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": "Home",
            "type": [Function],
          },
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "X",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "X",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Link
              className="menu-item"
              onClick={[Function]}
              replace={false}
              to="/login"
            >
              Login
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Login",
              "className": "menu-item",
              "onClick": [Function],
              "replace": false,
              "to": "/login",
            },
            "ref": null,
            "rendered": "Login",
            "type": [Function],
          },
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Link
              className="menu-item a-tag"
              onClick={[Function]}
              replace={false}
              to="/favorites"
            >
              Stalked: 
              <span>
                0
              </span>
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "Stalked: ",
                <span>
                  0
                </span>,
              ],
              "className": "menu-item a-tag",
              "onClick": [Function],
              "replace": false,
              "to": "/favorites",
            },
            "ref": null,
            "rendered": Array [
              "Stalked: ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": 0,
                },
                "ref": null,
                "rendered": 0,
                "type": "span",
              },
            ],
            "type": [Function],
          },
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
